proof-producer/tests/libs/output_artifacts/test_assignment_table_writer.cpp

```diff
@@ -0,0 +1,263 @@
+            Writer::write_text_assignment(out, table_, desc_, opts);
+            return out;
+        }
+
+        void read_and_check_text_header(std::istream& in) {
+            std::string expected_header = std::format("witnesses_size: {} public_inputs_size: {} constants_size: {} selectors_size: {} usable_rows_amount: {}", 
+                table_.witnesses_amount(), 
+                table_.public_inputs_amount(), 
+                table_.constants_amount(), 
+                table_.selectors_amount(), 
+                desc_.usable_rows_amount
+            );
+
+            std::string actual_header;
+            std::getline(in, actual_header);
+            EXPECT_EQ(expected_header, actual_header);
+        }
+
+        void read_separator(std::istream& in) {
+            std::string separator;
+            in >> separator;
+            EXPECT_EQ("|", separator);
+        }
+
+        // multiprectision number impl is quite library-specific (current boost impl works really strange with read from string)
+        // so for this test it is enough to compare string representations
+        std::string read_stringified_field(std::istream& in) {
+            std::string str;
+            in >> str;
+
+            constexpr auto is_zero = [](char c) { return c == '0'; };
+            
+            // merge 000000... to 0
+            if (std::all_of(str.begin(), str.end(), is_zero)) {
+                return "0";
+            }
+
+            // remove leading zeros inserted by setw
+            boost::algorithm::trim_left_if(str, is_zero);
+            return str;
+        }
+
+        void read_rest_of_line(std::istream& in) {
+            std::string rest;
+            std::getline(in, rest);
+            EXPECT_TRUE(std::all_of(rest.begin(), rest.end(), [] (char c) { return std::isspace(c); }));
+        }
+
+
+    protected:
+
+        constexpr static auto test_table_file_path = "./resources/assignment.tbl";
+
+        std::vector<std::uint8_t> table_bytes_;
+        AssignmentTable table_;
+        AssignmentTableDescription desc_{0,0,0,0};
+};
+
+TEST_F(AssignmentTableWriterTest, WriteBinaryAssignment) 
+{
+    std::stringstream out;
+    Writer::write_binary_assignment(out, table_, desc_);
+    out.flush();
+
+    ASSERT_EQ(out.tellp(), table_bytes_.size());
+    const auto written = out.rdbuf();
+    ASSERT_TRUE(std::memcmp(written->view().data(), table_bytes_.data(), table_bytes_.size()) == 0);
+}
+
+TEST_F(AssignmentTableWriterTest, WriteFullTextAssignment) 
+{
+    OutputArtifacts artifacts;
+    artifacts.write_full = true;
+
+    auto mem_stream = write_table_to_stringstream(artifacts);
+
+    read_and_check_text_header(mem_stream);
+    for (auto row = 0; row < desc_.usable_rows_amount; row++) {
+
+        for (auto col = 0; col < table_.witnesses_amount(); col++) {
+            auto expected = row < table_.witness_column_size(row) ? table_.witness(col)[row] : 0;
```

Comment:
may be use padded reference table? (no need the column size)

Reply:
yes, looks like redundant copypaste. Removed, thx